### PR TITLE
types: organize into folder

### PR DESCRIPTION
**Pull request by ignatiusmb**
Organize types into its own folder, this will make it easier to develop and reference types in deeply nested files, as well as add more unrelated types if necessary. This PR actions consists of

- organizing types and publishing the whole directory
- exposing only public types from `types/index.d.ts`
- escaping* relative path hell, reference like normal module

*note: there's still `runtime/client/types.d.ts` and `runtime/server/page/types.d.ts` which uses types from the root folder, but this could now be placed in the root `types` folder as well, it can (in theory) be referenced as `types/client` and `types/server`.

No tests or changeset should be needed as this should not change anything

### CODE REVIEW

**Review by benmccann — APPROVED** [review]
this looks very nice. thank you!

**Comment by dummdidumm** [comment]
While we're at it: could we also restructure the types internally? Right now they reference each other recursively which signals to me that there's something wrong. Could we change this up so that internal only references public but not the other way round? Everything in public that references internal is public, too, after all.

**Comment by ignatiusmb** [comment]
I tried my best to break it up even further. Also updated the paths used internally to import it explicitly (from `types/foo`) instead of from the public (`index.d.ts`) types directly, disabled it in tsconfig as well so it'll be consistent.

Although `internal.d.ts` is still a mess, I'm assuming the API aren't quite stable yet so I tried to minimize the changes there.

**Comment by dummdidumm** [comment]
I think these are good splits. I would suggest to revert the interface name changes though to not introduce a breaking change at the same time when we refactor the types, that should go into a different PR IMO.

**Review by dummdidumm — APPROVED** [review]

File: packages/kit/tsconfig.enforced.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
 		"module": "es2020",
 		"moduleResolution": "node",
 		"allowSyntheticDefaultImports": true,
+		"types": ["./types"],
```
Comment (dummdidumm) [inline packages/kit/tsconfig.enforced.json:11]: This closes off all other ambient types from appearing, which might fail in the future. I'm also not sure why this suddenly is a problem.

Comment (ignatiusmb) [reply packages/kit/tsconfig.enforced.json:11]: That might be a problem. Though, I really don't know why it suddenly decided to not detect the ambient types. The first 2 push seems to works just fine.

**Review by Rich-Harris — CHANGES_REQUESTED** [review]
Thank you, but can we change `types/server` to `types/endpoint` and `types/client` to `types/page`? The current division isn't quite accurate, since pages are rendered on the server as well as in the client.

**Comment by ignatiusmb** [comment]
Done!

I was following how it each types were imported from `runtime/client` and `runtime/server` respectively at first, but calling it `endpoint` and `page` seems better now.

**Review by Rich-Harris — APPROVED** [review]

**Comment by Rich-Harris** [comment]
excellent, thank you so much!

**Pull request merged**